zebra-script/src/lib.rs

```diff
@@ -56,85 +55,92 @@ impl From<zcash_script_error_t> for Error {
-        )
-    };
+/// A preprocessed Transction which can be used to verify scripts within said
+/// Transaction.
+pub struct PrecomputedTransaction {
+    precomputed: *mut std::ffi::c_void,
+}
 
-    if ret == 1 {
-        Ok(())
-    } else {
-        Err(Error::from(err))
+unsafe impl Send for PrecomputedTransaction {}
+
+impl PrecomputedTransaction {
+    /// Construct a `PrecomputedTransaction` from a `Transaction`.
+    pub fn new(transaction: &Transaction) -> Self {
+        let tx_to = transaction
+            .zcash_serialize_to_vec()
+            .expect("serialization into a vec is infallible");
+
+        let tx_to_ptr = tx_to.as_ptr();
+        let tx_to_len = tx_to.len() as u32;
+        let mut err = 0;
+
+        let precomputed = unsafe {
+            zcash_script::zcash_script_new_precomputed_tx(tx_to_ptr, tx_to_len, &mut err)
+        };
+
+        Self { precomputed }
+    }
+
+    /// Verify a script within a transaction given the corresponding
+    /// `transparent::Output` it is spending and the `ConsensusBranchId` of the block
+    /// containing the transaction.
+    ///
+    /// # Details
+    ///
+    /// input index corresponds to the index of the `TransparentInput` which in
+    /// `transaction` used to identify the `previous_output`
+    pub fn is_valid(
+        &self,
+        branch_id: ConsensusBranchId,
+        (input_index, previous_output): (u32, transparent::Output),
+    ) -> Result<(), Error> {
+        let transparent::Output { value, lock_script } = previous_output;
+        let script_pub_key: &[u8] = lock_script.0.as_ref();
+        let n_in = input_index as _;
+
+        let script_ptr = script_pub_key.as_ptr();
+        let script_len = script_pub_key.len();
+
+        let amount = value.into();
+        let flags = zcash_script::zcash_script_SCRIPT_FLAGS_VERIFY_P2SH
+            | zcash_script::zcash_script_SCRIPT_FLAGS_VERIFY_CHECKLOCKTIMEVERIFY;
+
+        let consensus_branch_id = branch_id.into();
+
+        let mut err = 0;
+
+        let ret = unsafe {
+            zcash_script::zcash_script_verify_precomputed(
+                self.precomputed,
+                n_in,
+                script_ptr,
+                script_len as u32,
+                amount,
+                #[cfg(not(windows))]
+                flags,
+                #[cfg(windows)]
+                flags.try_into().expect("why bindgen whyyy"),
+                consensus_branch_id,
+                &mut err,
+            )
+        };
+
+        if ret == 1 {
+            Ok(())
+        } else {
+            Err(Error::from(err))
+        }
     }
 }
 
-/// Verify a script within a transaction given the corresponding
-/// `transparent::Output` it is spending and the `ConsensusBranchId` of the block
-/// containing the transaction.
-///
-/// # Details
-///
-/// input index corresponds to the index of the `TransparentInput` which in
-/// `transaction` used to identify the `previous_output`
-pub fn is_valid(
```

Comment:
Removing this `fn` is a breaking change but because we almost certainly have nobody depending on the existing API and the fact that we've never published these crates I don't think we should worry about bumping the version number.

Reply:
I'm happy to go either way with this one.

Since we've tagged a release with the old version, it's good practice to bump the version for this change. But we didn't make any stability guarantees with that release, so maybe we don't need to do that.

Let's discuss at the Zebra sync?